### PR TITLE
content: Handle KaTeX math, with a rough preview

### PR DESCRIPTION
**Pull request by gnprice**
Fixes: #359

See discussion on visual design: https://chat.zulip.org/#narrow/stream/48-mobile/topic/TeX.20in.20Flutter.20app/near/1710839

Screenshots (of [these messages](https://chat.zulip.org/#narrow/stream/7-test-here/topic/math/near/1712027)):

| This PR | Zulip web |
| --- | --- |
| image | image |

### CODE REVIEW

**Comment by chrisbobbe** [comment]
I assume we want to resolve [this CZO thread](https://chat.zulip.org/#narrow/stream/48-mobile/topic/TeX.20in.20Flutter.20app/near/1713890) before I review this.

File: lib/model/content.dart [hunk 1]
```diff
@@ -792,6 +894,16 @@ class _ZulipContentParser {
     }
 
     if (localName == 'p' && classes.isEmpty) {
+      // Oddly, the way a math block gets encoded in Zulip HTML is inside a <p>.
+      if (element.nodes case [dom.Element(localName: 'span') && var child]) {
```
Comment (chrisbobbe) [inline lib/model/content.dart:898]: Hmm, I'm seeing a block-quoted math block with more than one child inside the `<p>`. The HTML has the expected span and then `<br>\n`.

https://chat.zulip.org/#narrow/stream/7-test-here/topic/Chris/near/1714716

`````
````quote
```math
1 + 1
```
````
`````

```
"<blockquote>\n<p><span class=\"katex-display\"><span class=\"katex\"><span class=\"katex-mathml\"><math xmlns=\"http://www.w3.org/1998/Math/MathML\" display=\"block\"><semantics><mrow><mn>1</mn><mo>+</mo><mn>1</mn></mrow><annotation encoding=\"application/x-tex\">1 + 1</annotation></semantics></math></span><span class=\"katex-html\" aria-hidden=\"true\"><span class=\"base\"><span class=\"strut\" style=\"height:0.7278em;vertical-align:-0.0833em;\"></span><span class=\"mord\">1</span><span class=\"mspace\" style=\"margin-right:0.2222em;\"></span><span class=\"mbin\">+</span><span class=\"mspace\" style=\"margin-right:0.2222em;\"></span></span><span class=\"base\"><span class=\"strut\" style=\"height:0.6444em;\"></span><span class=\"mord\">1</span></span></span></span></span><br>\n</p>\n</blockquote>"
```

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Here's a review from reading the code and some manual testing.

File: lib/model/content.dart [hunk 2]
```diff
@@ -532,6 +572,60 @@ class _ZulipContentParser {
...
+    if (grandchild.attributes['display'] != (block ? 'block' : null)) return null;
+    if (grandchild.namespaceUri != 'http://www.w3.org/1998/Math/MathML') return null;
+
+    if (grandchild.nodes.length != 1) return null;
+    final greatgrand = grandchild.nodes.single;
+    if (greatgrand is! dom.Element) return null;
+    if (greatgrand.localName != 'semantics') return null;
+
+    if (greatgrand.nodes.isEmpty) return null;
+    final child4 = greatgrand.nodes.last;
```
[the remark this answers is not in the document]
Comment (gnprice) [reply lib/model/content.dart:621]: Hmm, that's not the meaning I intended with this name. Definitely could use some changes to make it clearer, then!

I don't expect a fixed number of children here, of the `<semantics>` element. In particular if you look at your example below, or at the example in the test cases, it has two children — an `<mrow>` and then the `<annotation>` — but I figure it's probably not always an `<mrow>`, and may not always be just one sibling before the `<annotation>`.

Maybe a better name would be `descendant4`. It's the child of `greatgrand`, which is the great-grandchild of `katexElement`, the third generation down from `katexElement`; so this one is the fourth generation down.

The first name I started typing for it was `greatgreat`, but I figured that looks too similar to `greatgrand`. Could also say `greatgreatgrand`, I guess — or maybe `greatGrand` and `greatGreatGrand` would be the proper camel-casing.

File: lib/model/content.dart (hunk 1 above)
Comment (gnprice) [reply lib/model/content.dart:898]: Good find, thanks! I'll handle that, and also take this as a sign I should do some broader experimenting.

File: lib/model/content.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply lib/model/content.dart:621]: > Maybe a better name would be `descendant4`.

Sure, that sounds good.

**Comment by gnprice** [comment]
Revision pushed! PTAL.

In addition to the review items above, I changed the visual design to the one more people preferred in that thread. (Will update screenshots next.)

**Comment by chrisbobbe** [comment]
Hmm, looks like a test failure in the MathBlock content widget test.

**Comment by gnprice** [comment]
Oops, thanks — fixed.

**Comment by chrisbobbe** [comment]
Thanks, LGTM! Merging.

**Pull request merged**